src/app/src/main/kotlin/org/vaccineimpact/orderlyweb/PorcelainAPI.kt

```diff
@@ -0,0 +1,262 @@
+package org.vaccineimpact.orderlyweb
+
+import com.google.gson.Gson
+import com.google.gson.JsonElement
+import com.google.gson.JsonParser
+import com.google.gson.reflect.TypeToken
+import okhttp3.Headers.Companion.toHeaders
+import okhttp3.HttpUrl
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.MediaType.Companion.toMediaType
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.RequestBody
+import okhttp3.RequestBody.Companion.toRequestBody
+import org.json.JSONArray
+import org.json.JSONObject
+import org.vaccineimpact.orderlyweb.errors.PorcelainError
+
+interface PorcelainAPI
+{
+    @Throws(PorcelainError::class)
+    fun post(
+            url: String,
+            context: ActionContext,
+            rawRequest: Boolean = false,
+            transformResponse: Boolean = true
+    ): PorcelainResponse
+
+    @Throws(PorcelainError::class)
+    fun post(
+            url: String,
+            json: String,
+            queryParams: Map<String, String?>
+    ): PorcelainResponse
+
+    @Throws(PorcelainError::class)
+    fun get(url: String, context: ActionContext): PorcelainResponse
+
+    @Throws(PorcelainError::class)
+    fun get(url: String, queryParams: Map<String, String>): PorcelainResponse
+
+    @Throws(PorcelainError::class)
+    fun delete(url: String, context: ActionContext): PorcelainResponse
+
+    fun throwOnError(): PorcelainAPI
+}
+
+class PorcelainResponse(val bytes: ByteArray, val statusCode: Int)
+{
+    constructor(text: String, statusCode: Int) : this(text.toByteArray(), statusCode)
+
+    val text get() = String(bytes)
+
+    fun <T> data(klass: Class<T>): T
+    {
+        val data = parseJson(text)
+        return Serializer.instance.gson.fromJson(data, klass)
+    }
+
+    fun <T> listData(klass: Class<T>): List<T>
+    {
+        val data = parseJson(text)
+        val type = TypeToken.getParameterized(List::class.java, klass).type
+        return Serializer.instance.gson.fromJson(data, type)
+    }
+
+    private fun parseJson(jsonAsString: String): JsonElement
+    {
+        val element = JsonParser().parse(jsonAsString)
+        return element.asJsonObject["data"]
+    }
+}
+
+open class PorcelainAPIClient(
+        private val instanceName: String,
+        private val urlBase: String,
+        private val client: OkHttpClient = OkHttpClient()
+) : PorcelainAPI
+{
+    protected var throwOnError = false
+
+    private val standardHeaders = mapOf("Accept" to ContentTypes.json)
+
+    override fun throwOnError(): PorcelainAPI
+    {
+        return apply {
+            throwOnError = true
+        }
+    }
+
+    override fun get(url: String, context: ActionContext): PorcelainResponse
+    {
+        val request = Request.Builder()
+                .url(buildFullUrl(url, context.queryString()))
+                .headers(standardHeaders.toHeaders())
+                .build()
+
+        client.newCall(request).execute().use {
```

Comment:
I lied, there is one logic change. Previously this and all other methods were using 

```val response = client.newCall(request).execute()```

leading to leaked connections. Have changed every method to use a `use` block instead, ensuring connections get closed.